lighthouse-core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js

```diff
@@ -20,162 +19,184 @@
-  return new Promise((resolve, reject) => {
-    try {
-      const tagList = [...document.querySelectorAll('link, head script[src]')]
-        .filter(tag => {
-          if (tag.tagName === 'SCRIPT') {
-            const scriptTag = /** @type {HTMLScriptElement} */ (tag);
-            return (
-              !scriptTag.hasAttribute('async') &&
-              !scriptTag.hasAttribute('defer') &&
-              !/^data:/.test(scriptTag.src) &&
-              !/^blob:/.test(scriptTag.src) &&
-              scriptTag.getAttribute('type') !== 'module'
-            );
-          } else if (tag.tagName === 'LINK') {
-            // Filter stylesheet/HTML imports that block rendering.
-            // https://www.igvita.com/2012/06/14/debunking-responsive-css-performance-myths/
-            // https://www.w3.org/TR/html-imports/#dfn-import-async-attribute
-            const linkTag = /** @type {HTMLLinkElement} */ (tag);
-            const blockingStylesheet = linkTag.rel === 'stylesheet' &&
-              window.matchMedia(linkTag.media).matches && !linkTag.disabled;
-            const blockingImport = linkTag.rel === 'import' && !linkTag.hasAttribute('async');
-            return blockingStylesheet || blockingImport;
-          }
-
-          return false;
-        })
-        .map(tag => {
-          return {
-            tagName: tag.tagName,
-            url: tag.tagName === 'LINK' ? tag.href : tag.src,
-            src: tag.src,
-            href: tag.href,
-            rel: tag.rel,
-            media: tag.media,
-            disabled: tag.disabled,
-            mediaChanges: window.___linkMediaChanges.filter(item => item.href === tag.href),
-          };
-        });
-      resolve(tagList);
-    } catch (e) {
-      const friendly = 'Unable to gather Scripts/Stylesheets/HTML Imports on the page';
-      reject(new Error(`${friendly}: ${e.message}`));
-    }
-  });
+async function collectTagsThatBlockFirstPaint() {
+  /** @type {Array<MediaChange>} */
+  // @ts-expect-error - `___linkMediaChanges` created in `installMediaListener`.
+  const linkMediaChanges = window.___linkMediaChanges;
+
+  try {
+    const linkTags = [...document.querySelectorAll('link')]
+      .filter(/** @return {tag is HTMLLinkElement} */ tag => {
+        if (tag.tagName !== 'LINK') return false;
+
+        // Filter stylesheet/HTML imports that block rendering.
+        // https://www.igvita.com/2012/06/14/debunking-responsive-css-performance-myths/
+        // https://www.w3.org/TR/html-imports/#dfn-import-async-attribute
+        const linkTag = /** @type {HTMLLinkElement} */ (tag);
+        const blockingStylesheet = linkTag.rel === 'stylesheet' &&
+          window.matchMedia(linkTag.media).matches && !linkTag.disabled;
+        const blockingImport = linkTag.rel === 'import' && !linkTag.hasAttribute('async');
+        return blockingStylesheet || blockingImport;
+      })
+      .map(tag => {
+        return {
+          tagName: /** @type {'LINK'} */ ('LINK'),
```

Comment:
nit: This cast seems odd. Instead, I think you can enforce type `Array<LinkTag>` on `linkTags` at L61.

Reply:
> nit: This cast seems odd. Instead, I think you can enforce type `Array<LinkTag>` on `linkTags` at L61.

since we don't have `as const` in jsdoc, you get used to it :) (and `as const` itself is awkward, but widening heuristics are hard).

In this case, though, the container type is simple so that change sounds good 👍